quedex_api/user_stream.py

```diff
@@ -0,0 +1,302 @@
+      {
+        "client_order_id": <integer id of the order to cancel>,
+      }
+    """
+    if not self._initialized:
+      raise Exception('UserStream not initialized, wait until UserStreamListener.on_read is called')
+    cancel_order_command['type'] = 'cancel_order'
+    self._encrypt_send(self._set_nonce_account_id(cancel_order_command))
+
+  def modify_order(self, modify_order_command):
+    """
+    :param modify_order_command: a dict of the following format:
+      {
+        "client_order_id": <integer id of the order to modify>,
+        "new_limit_price": "<decimal as string>",
+        "new_quantity": <integer>,
+      }
+    """
+    if not self._initialized:
+      raise Exception('UserStream not initialized, wait until UserStreamListener.on_read is called')
+    modify_order_command['type'] = 'modify_order'
+    self._encrypt_send(self._set_nonce_account_id(modify_order_command))
+
+  def batch(self, order_commands):
+    """
+    :param order_commands: a list with the number of commands where the following are possible:
+     [
+      {
+        "type": "place_order",
+        // for the rest of the fields see place_order method
+      },
+      {
+        "type": "cancel_order",
+        // for the rest of the fields see cancel_order method
+      },
+      {
+        "type": "modify_order",
+        // for the rest of the fields see modify_order method
+      },
+      ...
+     ]
+    """
+    if not self._initialized:
+      raise Exception('UserStream not initialized, wait until UserStreamListener.on_read is called')
+    for command in order_commands:
+      self._set_nonce_account_id(command)
+    self._encrypt_send({
+      'type': 'batch',
+      'batch': order_commands,
+    })
+
+  def initialize(self):
+    self._encrypt_send({
+      'type': 'get_last_nonce',
+      'nonce_group': self._nonce_group,
+      'account_id': self._account_id
+    })
+
+  def on_message(self, message_wrapper_str):
+    try:
+      if message_wrapper_str == 'keepalive':
+        return
+
+      message_wrapper = json.loads(message_wrapper_str)
+
+      if message_wrapper['type'] == 'error':
+        # error_code == maintenance accompanies exchange engine going down for maintenance which
+        # causes graceful disconnect of the WebSocket, handled by MarketStreamListener.on_disconnect
+        if message_wrapper['error_code'] != 'maintenance':
+          self.on_error(Exception('WebSocket error: ' + message_wrapper['error_code']))
+        return
+
+      entity = self._decrypt(message_wrapper['data'])
+
+      if entity['type'] == 'last_nonce':
+        self._nonce = entity['last_nonce']
+        self._encrypt_send(self._set_nonce_account_id({'type': 'subscribe'}))
+        return
+      elif entity['type'] == 'subscribed':
+        self._initialized = True
+        self._call_listeners('on_ready')
+        return
+
+      self._call_listeners('on_' + entity['type'], entity)
+      self._call_listeners('on_message', entity)
```

Comment:
I think that more general tasks are to be done earlier (e.g logging each message), but now I think that we may actually need two methods :)